Replace nonce queue with single-threaded implementation

theodorebugnet opened the pull request:
# Description

Having the nonce buffers be single-threaded should make it significantly easier to reason about.

Integration tests pass. Unit tests from the old implementation will be ported in follow-up work.

The main `run()` function is also quite huge and badly needs refactoring, this will be done in immediate follow-up work.

- [ ] I have updated `CHANGELOG.md` with a new entry if my PR makes any breaking changes or fixes a bug. If my PR removes a feature or changes its behavior, I provide help for users on how to migrate to the new behavior.
- [ ] I have carefully reviewed all my `Cargo.toml` changes before opening the PRs. (Are all new dependencies necessary? Is any module dependency leaked into the full-node (hint: it shouldn't)?)

## Linked Issues
- Fixes # (issue, if applicable)
- Related to # (issue) 

## Testing
Describe how these changes were tested. If you've added new features, have you added unit tests?

## Docs
Describe where this code is documented. If it changes a documented interface, have the docs been updated?

Rendered docs are available at `https://sovlabs-ci-rustdoc-artifacts.us-east-1.amazonaws.com/<BRANCH_NAME>/index.html`

bkolad committed:
Inc gas_limit

bkolad committed:
Timing oracle

LogvinovLeon committed:
Extract metrics tracking

LogvinovLeon committed:
Simplify handlers

LogvinovLeon committed:
Create struct to dedupe handlers where clauses

LogvinovLeon committed:
Move signing methods into Handlers struct

LogvinovLeon committed:
Add eth_sendRawTransactionSync

LogvinovLeon committed:
Extract receipt getting

preston-evans98 committed:
Fix foundry

LogvinovLeon committed:
Use RpcResult

LogvinovLeon committed:
Implement tx timeout on eth_send_raw_transaction_sync

LogvinovLeon committed:
Don't swallow parsing errors

LogvinovLeon committed:
Fix timeout message

preston-evans98 committed:
Merge branch 'leo/eth-send-raw-transaction-sync' of github.com:Sovereign-Labs/sovereign-sdk into preston/relay-testnet

preston-evans98 committed:
revert nomt

preston-evans98 committed:
Switch to nomt without page elision

preston-evans98 committed:
Fill zero hash for pending receipts

theodorebugnet committed:
Add extra information when the sequencer nonce queues reject a tx

theodorebugnet committed:
fmt

theodorebugnet committed:
lockfile needed changing for some reason

preston-evans98 committed:
Improve invalid nonce error message

theodorebugnet committed:
Fixes a bug in sequencer nonce queue when a transaction is rejected. (#2131)

Fixes a bug in sequencer nonce queue when a transaction is rejected.
The nonce queue optimistically keeps track of nonces submitted for execution, since the actual API state only updates after a transaction is persisted to the DB; tracking the executing nonce locally allows queueing up transactions with consecutive nonces without waiting for the DB latency.
The problem was that this locally tracked "last submitted for execution" nonce did not get reset if said submitted transaction actually got rejected. This had two consequences: first and less importantly, transactions would not be evicted even when no longer having all pre-requisites. This could've had a theoretical memory impact if someone filled their queue with transactions and then sent a rejected transaction at the head, resulting in the sequencer keeping the entire queue in memory forever. The more important issue was that the sequencer would then consider transactions with the next nonce valid and would submit them for execution. They would then always get rejected due to the user's nonce, of course, not having been actually incremented by the previous (rejected) transaction. This would spiral into a vicious cycle of invalidating and rejecting large swathes of transactions and causing terrible user experience.

theodorebugnet committed:
Bump sequencer queue timeout, sendrawtransaction_sync timeout, and add a debug log to all ethereum transactions ingested (#2136)

theodorebugnet committed:
Initial implementation. Integration tests pass, needs refactoring and unit tests

theodorebugnet committed:
Fix queue cleanup. Small non-persisted adjustment. Initial unit test porting.

theodorebugnet committed:
Remove non-working unit tests for now.

theodorebugnet committed:
Remove panics from queue for now

theodorebugnet committed:
Revert "Remove non-working unit tests for now."

This reverts commit 18757f2e67c8bb05d3b085b803097a419c0af78a.

preston-evans98 commented on crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs:
```diff
@@ -0,0 +1,800 @@
+
+    fn mark_inflight_execution_succeeded(&mut self, tx_nonce_check: u64) {
+        if !self.has_in_flight {
+            tracing::error!("Sequencer nonce buffer: non-persisted tracking: attempted to mark executed tx successful when has_in_flight is false!");
```
This seems like it should probably be at least a debug assertion - otherwise, we won't catch bugs in our tests.

preston-evans98 commented on crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs:
```diff
@@ -0,0 +1,800 @@
+            .user_nonce()
+            .is_some_and(|user_nonce| tx_nonce_check != user_nonce)
+        {
+            tracing::error!("Sequencer nonce buffer: non-persisted tracking: attempted to execute nonce that does not match tracked next nonce!");
```
This should probably be a debug assertion as well. (Or, even better, a real assertion)

theodorebugnet replied:
I'm shying away from making it a real assertion because if it fails, the entire sequencer will crash, while the only error will have been slightly less reliable nonce queueing behaviour. IMO it's better to have extra rejections from the sequencer than complete rollup crashes. Of course it'd be good to have strict correctness guarantees but not at the cost of bringing down customer rollups.
LMK if you disagree though

preston-evans98 commented on crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs:
```diff
@@ -0,0 +1,800 @@
+
+    async fn run(&mut self) {
+        while let Some(input) = self.buffer_input.recv().await {
+            println!("Nonce buffer task processing input {input:?}");
```
nit: This look like leftover debug output

preston-evans98 commented on crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs:
```diff
@@ -0,0 +1,800 @@
+                        self.execution_backend
+                            .get_current_nonce_for_user(&credential_id)
+                    });
+                    println!("User nonce during NewTx: {user_nonce}");
```
nit: leftover println

preston-evans98 commented on crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs:
```diff
@@ -0,0 +1,800 @@
+                                    InvalidNonceReason::Replaced,
+                                ));
+                            }
+                            self.schedule_timeout(credential_id, tx_nonce, tx_hash);
```
I think there's an edge case here where the user accidentally submits the exact same tx (with the same nonce) twice. Then, we'll have multiple timeouts scheduled for the same job. Two easy fixes: 
1. compare the hashes of new and old_tx. Ignore new_tx if it's got the same hash 
2. Add a oneshot to the timeout job which gets dropped when we execute or evict a txs. Then timeouts will get cleaned up immediately.

theodorebugnet replied:
I'm not sure it would've caused issues, but checking the old_tx hash doesn't add extra complexity and eliminates the issue so I'll go ahead and add it :+1:

preston-evans98 commented on crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs:
```diff
@@ -0,0 +1,800 @@
+                            self.schedule_timeout(credential_id, tx_nonce, tx_hash);
+                        }
+                        Action::Execute => {
+                            println!("Determined action: Execute (for nonce {tx_nonce})");
```
nit: Leftover println

preston-evans98 commented on crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs:
```diff
@@ -0,0 +1,800 @@
+                            });
+                        }
+                        Action::Reject => {
+                            println!("Determined action: Execute (for nonce {tx_nonce})");
```
nit: Leftover println

preston-evans98 commented on crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs:
```diff
@@ -0,0 +1,800 @@
+                        .non_persisted
+                        .user_nonce()
+                        .expect("We just marked the nonce on the queue, so it cannot be None");
+                    println!("\nAfter execution, next nonce being queued is {user_nonce}");
```
nit: leftover println

preston-evans98 commented on crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs:
```diff
@@ -0,0 +1,800 @@
+                                // Stale transaction in queue - evict and ignore.
+                                // This should not normally happen either, but we handle it to avoid a deadlock
+                                // if it does happen for any reason.
+                                head_entry.remove();
```
nit: Let's log an error here so that we can detect bugs

preston-evans98 replied:
Even better, let's add a debug assertion

preston-evans98 commented on crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs:
```diff
@@ -0,0 +1,800 @@
+                    if entry.get().txs.is_empty() {
+                        let real_nonce = self
+                            .execution_backend
+                            .get_current_nonce_for_user(&credential_id);
```
nit: I think we can avoid the problems of asynchrony between the backend and the tx queue here by just putting the executed nonce into the `TxPersisted` message. We already have access to it when we create the message, and it'll also let us save the state read

theodorebugnet replied:
I had it that way at first but had decided it'd be easier to just check the real nonce once - if txs are inserted in batches into the DB, the TxPersisted messages could come all at once in arbitrary order, and just reading the state lets us clean up the queue once and the rest will be no-ops. Doing it based on the message does save the state read though you're right.

But can you clarify what asynchrony you're worried about here? I don't see any issues other than the state read optimisation, so I wanna make sure I'm not missing anything

preston-evans98 replied:
The asynchrony is just the fact that an arbitrary number of other txs might have been executed and persisted to the api state before we process this message; on the other hand, the message in the queue are guaranteed to be processed in order.

preston-evans98 commented on crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs:
```diff
@@ -0,0 +1,800 @@
+                    tx_hash,
+                } => {
+                    let queue = self.buffers.entry(credential_id).or_default();
+                    let state_nonce = self
```
nit: we can skip the state read here if `user_nonce_to_use_as_prerequisite_start`. Let's put it in an `unwrap_or_else`

preston-evans98 commented on crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs:
```diff
@@ -0,0 +1,800 @@
+                        .user_nonce_to_use_as_prerequisite_start()
+                        .unwrap_or(state_nonce);
+                    if queue.has_contiguity_between(user_nonce_for_prerequisites, tx_nonce) {
+                        self.schedule_timeout(credential_id, tx_nonce, tx_hash);
```
If we have contiguity we'll never actually time out the transaction, right? We'll just keep hitting this condition and scheduling a new timeout 20 seconds in the future; which means that (under load) the response time for these transactions is unbounded. Maybe we should pop them anyway?

theodorebugnet replied:
Correct. That's an intentional feature of the design - also kind of meant to be used with a much shorter timeout than 20s. The response time is bounded by how long it can possibly take to execute the rest the pre-requisite transactions.

The behaviour under load is a good point. We could add a max retry counter, so even if pre-reqs are satisfied, if the tx still isn't executed after N timeouts it's evicted anyway. Removing the contiguity check entirely would force all transactions to execute within the timeout window which seems strict.

preston-evans98 commented on crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs:
```diff
@@ -0,0 +1,800 @@
+) -> TransactionReceiverResult<S, Rt> {
+    let was_queued_msg = format!("The sequencer queued the transaction for reordering {} ms ago, when the user's nonce was {nonce_when_queued}.", instant_queued.elapsed().as_millis());
+    let queue_error_msg = match queue_rejection_reason {
+        InvalidNonceReason::Invalid => "The sequencer did not attempt to queue the transaction as it was not within valid queue limits (either in the past, or beyond the max limit).".to_string(),
```
nit: Let's include the tx_nonce, the account_nonce, and the max limit

theodorebugnet replied:
The tx_nonce and account_nonce are part of the full error message below ("Tx bad nonce..."). I'll add the max limit :+1:

theodorebugnet replied:
Actually adding the max_limit is quite annoying because it's a property of the NonceBufferTask, which means either
a) adding yet another parameter to the function, which already takes a lot and also complicates its callsite in the NonceBufferInputSender (which would probably need to take it as a param from the sequencer itself), or
b) making it a member function of NonceBufferTask which again causes issues for NonceBufferInputSender.

It's doable but IMO the information is not very useful - we know the config and we know the `expected_nonce`, and the max nonce will always be `expected_nonce + max_future_nonce_delta` which can always be had from the instance's config.toml.

Let me know if you have pushback, otherwise I'll skip this for now to avoid the extra complexity from passing it around.

preston-evans98 replied:
Fine by me!

preston-evans98 commented on crates/full-node/sov-sequencer/src/preferred/nonce_buffer_task.rs:
```diff
@@ -0,0 +1,800 @@
+    let was_queued_msg = format!("The sequencer queued the transaction for reordering {} ms ago, when the user's nonce was {nonce_when_queued}.", instant_queued.elapsed().as_millis());
+    let queue_error_msg = match queue_rejection_reason {
+        InvalidNonceReason::Invalid => "The sequencer did not attempt to queue the transaction as it was not within valid queue limits (either in the past, or beyond the max limit).".to_string(),
+        InvalidNonceReason::Timeout => format!("{was_queued_msg} In that time, the sequencer did not receive all the transactions leading up to this tx's nonce, so it has timed out and is being evicted from the queue."),
```
nit: let's change "did not receive" to "did not accept" so that we have flexibility to reject txs even if their prereqs have been enqueued.

theodorebugnet committed:
Refactor and convert most unit tests to new implementation

theodorebugnet committed:
Replacement unit tests added

theodorebugnet committed:
PR comments: disable pre-req checks on timeout, add debug assertions

theodorebugnet committed:
fmt and prints

preston-evans98 committed:
Squash single-threaded nonce queue impl

theodorebugnet committed:
Fix queue behaviour on sequencer resync.

The queue assumes that all transactions go through it. On resync, this isn't true, and the API state will increase independently of the queue. We detect this now by a) checking the original_tx_queue_id and b) detecting SequencerNotReady errors; both of these cases invalidate the entire nonce queue and cause a wipe.

theodorebugnet committed:
Fix queue behaviour on sequencer resync.

The queue assumes that all transactions go through it. On resync, this isn't true, and the API state will increase independently of the queue. We detect this now by a) checking the original_tx_queue_id and b) detecting SequencerNotReady errors; both of these cases invalidate the entire nonce queue and cause a wipe.

theodorebugnet committed:
PR comments

preston-evans98 committed:
Allow reverted txs onchain via config (#2141)

---------

Co-authored-by: Theodore Bugnet <theodore@sovlabs.io>

theodorebugnet committed:
Explicitly reject transactions on queue wipe, and restore unknown reason error when oneshot is dropped

theodorebugnet committed:
Merge branch 'theodore/single-threaded-nonce-queue-relay' of github.com:Sovereign-Labs/sovereign-sdk into theodore/single-threaded-nonce-queue

theodorebugnet committed:
Merge branch 'dev' of github.com:Sovereign-Labs/sovereign-sdk into theodore/single-threaded-nonce-queue

theodorebugnet committed:
Fix spurious config changes

preston-evans98 committed:
Merge branch 'dev' of github.com:Sovereign-Labs/sovereign-sdk into theodore/single-threaded-nonce-queue

preston-evans98 committed:
Fix types for latest dev

preston-evans98 committed:
fmt

theodorebugnet committed:
fix tests

preston-evans98 committed:
Merge branch 'dev' of github.com:Sovereign-Labs/sovereign-sdk into theodore/single-threaded-nonce-queue

preston-evans98 committed:
Group ip and credential in nonce queue

preston-evans98 committed:
Unflake bank tests

preston-evans98 committed:
Update changelog

preston-evans98 committed:
Merge branch 'dev' into theodore/single-threaded-nonce-queue
